Add `testError()` syntax sugar for the self-documenting general purpose errors unit tests

mokagio opened the pull request:
**~~Good idea or bad idea?~~** Update: Thank you all for the feedback. Promoted to good idea 😄 

I find having to think about what error to create in a test adds unnecessary friction. In some project, I used to create a dedicated `struct TestError: Error` type, but @crazytonyli pointed out that using an `NSError` might be best. It's definitely one less type.

Enter, `NSError.testInstance()` a method to use in the tests to generate a hassle-free, consistent, yet customizable placeholder error.

~~What do you think? Does this remove noise from the tests? Or is it the need to know about this convention a point of friction?~~

On top of that, to make the usage smoother, the unit tests target has a `testError()` that can be used in any place where `Error` is required.

---

## Regression Notes

1. Potential unintended areas of impact – N.A.
2. What I did to test those areas of impact (or what existing automated tests I relied on) – N.A.
3. What automated tests I added (or what prevented me from doing so) – N.A.

---

**PR submission checklist:**

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes. **N.A.**
- [x] I have considered adding accessibility improvements for my changes. **N.A.**
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary. **N.A.**

**UI changes testing checklist:** Not a UI PR.

mokagio committed:
Define `.testError()` `static` builder on `Error`

The idea is to use this to create any kind of test error in the tests.
It should replace at least three different implementation currently in
use in the tests.

mokagio committed:
Move `testError()` out of single test file

This required making it an `NSError` extensions, which I'm not too sure
about.

mokagio committed:
Use `NSError.testError()` through part of the test suite

mokagio committed:
Add note about `Error` extension limitation in tests

mokagio committed:
fixup! Use `NSError.testError()` through part of the test suite

mokagio committed:
Remove now unused `TestError` `struct`

mokagio committed:
Redefine `TestError`, this time as an `NSError` `typealias`

mokagio committed:
Replace some `NSError.testError()` calls with `TestError()`

mokagio committed:
Move away from `TestError` `typealias` because calling it crashes

The message we got when calling `TestError()` was that it was calling
`NSError init` which is not valid.

alpavanoglu commented:
It makes sense @mokagio 

We could also use the Builder format for Model creation. Right now we have examples like PageBuilder in the app. A more general way would be to have a protocol like
```Swift
protocol EntitiyBuilder {
  var dictionary: [String: Any?] { get }
}
```
And we could modify the dictionary in the with functions and return self. Of course this protocol wouldn't be for the Error but for Codable Models.

What I like about this pattern is one can ignore all the details if they simply want an instance. Or modify only the relevant pieces and see a list of all modification functions by simply typing .with

It essentially would achieve the same goal I think but would be scalable to every other model/error with the same syntax.

In any case, I think this is a helpful utility to have in the codebase.

kean commented:
I usually use `NSError` or `URLError(code: .unknown)` in tests when I don't care about the type of error it throws. But sometimes, it's nice to have something (easily) equatable. With `NSError`, you can check the domain and/or the code, which is also OK.

mokagio committed:
Add global free function `testError()` to simplify errors in tests

mokagio committed:
Extract `NSError` `extension`s for tests in dedicated file

mokagio committed:
Replace `NSError.testInstance` usages with cleaner `testError`

Hat tip @jkmassel for making think outside the class box and use a free
function.

mokagio commented:
@kean @alpavanoglu thank you for the feedback on this 🙌 

Regarding the having builders, that's definitely something we should invest in. As a matter of fact, my next task was to ensure the tests that create `Post` or `Account` always use the respective builders:

https://github.com/wordpress-mobile/WordPress-iOS/blob/aee37b66acb268e64b801bdefecb752ed9b21828/WordPress/WordPressTest/AccountBuilder.swift

https://github.com/wordpress-mobile/WordPress-iOS/blob/aee37b66acb268e64b801bdefecb752ed9b21828/WordPress/WordPressTest/PostBuilder.swift

@alpavanoglu Could you elaborate on how you'd envision using the protocol you mentioned:

```swift
protocol EntitiyBuilder {
  var dictionary: [String: Any?] { get }
}
```

crazytonyli commented on WordPress/WordPressTest/TestError.swift:
```diff
@@ -1,8 +1,28 @@
+        code: Int = 1
+    ) -> Self {
+        // This would be discouraged if it wasn't for the facts that we know that NSError converts to Error.
+        NSError.testInstance(description: description, domain: domain, code: code) as! Self
```
This force cast may crash I think? Because `Self` is not always `NSError`? Like, calling `CancellationError.testError` probably is a crash?

mokagio replied:
Possible. I didn't try it actually and opted for removing the `Error` extension that was, after all, unused.

mokagio committed:
Remove unused `Error` extension for tests
